File: 867.transpose-matrix.cpp
/*
 * @lc app=leetcode id=867 lang=cpp
 *
 * [867] Transpose Matrix
 */

#include <bits/stdc++.h>
#include "lassert/lassert.h"

using namespace std;

// @lc code=start
class Solution {
 public:
  vector<vector<int>> transpose(vector<vector<int>>& matrix) {
    vector<vector<int>> ret;
    int width = matrix[0].size();
    int height = matrix.size();
    ret.resize(width);
    for (int x = 0; x < width; x++) {
      ret[x].resize(height);
      for (int y = 0; y < height; y++) {
        ret[x][y] = matrix[y][x];
      }
    }
    return ret;
  }
};
// @lc code=end

int main() {
  Solution s;
  vector<vector<int>> matrix, exp;

  matrix = {{1, 2, 3}, {4, 5, 6}, {7, 8, 9}};
  exp = {{1, 4, 7}, {2, 5, 8}, {3, 6, 9}};
  assert(s.transpose(matrix) == exp);

  matrix = {{1, 2, 3}, {4, 5, 6}};
  exp = {{1, 4}, {2, 5}, {3, 6}};
  assert(s.transpose(matrix) == exp);
  return 0;
}

File: 867.transpose-matrix.rs
/*
 * @lc app=leetcode id=867 lang=rust
 *
 * [867] Transpose Matrix
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn transpose(matrix: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
        let mut ret = Vec::<Vec<i32>>::new();
        let width = matrix[0].len();
        let height = matrix.len();
        ret.resize(width, vec![]);
        for x in 0..width {
            ret[x].resize(height, 0);
            for y in 0..height {
                ret[x][y] = matrix[y][x];
            }
        }
        ret
    }
}
// @lc code=end

#[test]
fn test1() {
    assert_eq!(
        Solution::transpose(vec![vec![1, 2, 3], vec![4, 5, 6], vec![7, 8, 9]]),
        vec![vec![1, 4, 7], vec![2, 5, 8], vec![3, 6, 9]]
    );
}

#[test]
fn test2() {
    assert_eq!(
        Solution::transpose(vec![vec![1, 2, 3], vec![4, 5, 6]]),
        vec![vec![1, 4], vec![2, 5], vec![3, 6]]
    );
}
